Add 2021 year in review post (#266)
trinodb/trino.io: _posts/2021-12-31-trino-2021-a-year-of-growth.md

> ---

Reviewer: I would add a "by the numbers" section somewhere that just shows a highlight of all the quantifiable metrics: slack members, commits, TCB episodes, youtube/twitter followers, releases, etc.

layout: post
title:  "Trino 2021 Wrapped: A Year of Growth"
author: Brian Olsen, Martin Traverso, Manfred Moser
excerpt_separator: <!--more-->
---

> As we reflect on Trino’s 2021 wrapped, one thing stands out. Compared to 

Author: Ah fair enough. I was trying to make a play on words of [Spotify's wrapped ](https://www.spotify.com/us/wrapped/) playlist that has become a year-end trend but it's not as well known as I thought.

previous years we have seen even further accelerated, tremendous growth. Yes,
this is what all these year-in-retrospect blog posts say, but this has some 
special significance to it. This week marked the one-year anniversary since the 
project [dropped the Presto name and moved to the Trino name]({{site.url}}/blog/2020/12/27/announcing-trino.html).

## Renaming from PrestoSQL to Trino

This was a gut wrenching decision for many in the community, who had grown with
the project from early days. It also meant that the community would have to
build affinity towards the new name from scratch. We weren't sure what to expect
when we announced the rename. What followed, was an outpouring of support and
shock from the larger developer community. 

<p align="center">
 <a href="https://twitter.com/trinodb/status/1343330429684703232?s=20" target="_blank">
   <img align="center" width="100%" src="/assets/blog/2021-review/tweets.png"/>
 </a>
</p>

<!--more-->

The community immediately got to work. The project had to change the namespace
practically overnight from the `io.prestosql` namespace to `io.trino` and

> a [migration blog post]({{site.url}}/blog/2021/01/04/migrating-from-prestosql-to-trino.html)

Reviewer: This needs to be exposed more prominently. Specifically, where to find the code, slack, docker images, etc. Some people are still getting confused or lost.

* [Metabase](https://github.com/metabase/metabase/issues/17532)
* and so much more...

Not only was the community excited and supportive about the brand change, but
particularly they were all loving the new mascot. Our adorable bunny was soon 
after [named Commander Bun Bun by the community](/episodes/10.html).

<p align="center">
 <a href="https://twitter.com/jtannady/status/1346888143459545092" target="_blank">
   <img align="center" width="50%" src="/assets/blog/2021-review/cbb.png"/>
 </a>
</p>

Immediately after the announcement, the [Trino GitHub repository](https://github.com/trinodb/trino)
started trending in number of stargazers. Up until this point, the PrestoSQL
GitHub repository had only amassed 1,600 stargazers in the two years since it 
had split from the PrestoDB repository. However, within four months after the 
renaming, the number of stargazers had doubled. GitHub stars, issues, pull 
requests and commits started growing at a new trajectory.

<p align="center">
 <a href="https://twitter.com/bitsondatadev/status/1344028682126565381" target="_blank">
   <img align="center" width="50%" src="/assets/blog/2021-review/trending.png"/>
 </a>
</p>

## Community growth

> While we are on the subject, lets talk about the health of the Trino community.

Reviewer: let's

> As mentioned the GitHub stargazers trajectory skyrocketed. At the time of 

Reviewer: "As mentioned above". 

However, I think this paragraph is too repetitive. The numbers were just discussed in the section above. I would either some of the details that weren't discussed to the previous section and just get rid of this paragraph. Maybe just keep the intro sentence.

writing, we just hit 4,600 stargazers on GitHub. This means, we have added over
3,000 stargazers in the last year, a 187% increase.

> Aside from GitHub, we have seen a 47% growth of [the Trino Twitter](https://twitter.com/trinodb) 

Reviewer: "Aside from GitHub, [...]": aside from github what? This fragment is incomplete. Once you fix the paragraph below, it might be easier to restructure.

followers this year. [The Trino Slack community](https://trino.io/slack.html),
where a large amount of troubleshooting and development discussions occur, saw a

> 75% growth nearing 6,000 members. Finally, [the Trino YouTube channel](https://www.youtube.com/c/TrinoDB)

Reviewer: "growth, nearing [...]" (missing comma)

has seen an impressive 280% growth in subscribers as the [Trino Community Broadcast](/broadcast/)

> has grown in the number of helpful videos for community members.

Reviewer: > has grown in the number of helpful videos for community members.

This sounds a bit meh. Maybe talk about the number of episodes we published on various topics (with some examples)

Author: Agreed.

A lot of the increase on this channel was due to the [Trino Community Broadcast](/broadcast/), 
that brought users and contributors from the community to cover 23 episodes
about the following topics:

* Trino ecosystem (dbt, Amundsen, Debezium, Superset) - 7
* Trino project (Renaming Trino, Intro to Trino, )  - 4
* Trino connectors (Iceberg, Druid, Pinot) - 4
* Trino internals (Distributed Hash-Joins, Dynamic Filtering, Views) - 4
* Trino on Kubernetes (Trinetes series) - 2
* Trino users (LinkedIn, Resurface) - 2

While stargazers, subscribers, and followers tell the story of the growing 
awareness of the Trino project with the new name, what about the actual rate of
development on the project?

At the start of the year, there were 21,924 commits. This year, we pushed 3,679 

> commits to the repository,sitting at over 25,600 now. Looking at the graph, this

Reviewer: space after comma: "repository, sitting"

> keeps us pretty consistent with 2020s throughput.

Reviewer: 2020's

<p align="center">
 <img align="center" width="75%" src="/assets/blog/2021-review/commits.png"/>
</p>

With the project's trajectory displayed in numbers, let's examine the top 
features that landed in Trino this year.

> ## Features

Reviewer: HTTP Event listener: https://trino.io/docs/current/admin/event-listeners-http.html

Reviewer: Custom connector metrics

Reviewer: Transparent query retries

Reviewer: Column and table lineage tracking in QueryCompletedEvent

Here's a high-level list of the most exciting features that made their way into
Trino in 2021. For details and to keep up you can check out the [release notes]({{site.url}}/docs/current/release.html).

> ### Improvements to ANSI SQL support

Reviewer: Add links to related blog posts, documentation or roadmap issues wherever possible to give a chance to readers to find more information.

> ### Improvements to ANSI SQL support

Reviewer: Change

> `Improvements to ANSI SQL support`

to

> `SQL language improvements`

Not all the improvements are part of the SQL standard (e.g, DENY, IN <catalog>, SHOW STATS, ALTER MATERIALIZED VIEW, etc)

Reviewer: Also, let's group the items based on focus area (e.g., list all querying related improvements first -- MATCH_RECOGNIZE, RANGE/ROWS, etc --, followed by functions, metadata management and security)

> ### Improvements to ANSI SQL support

Reviewer: WINDOW clause: https://trino.io/docs/current/sql/select.html#window-clause

> ### Improvements to ANSI SQL support

Reviewer: to/from GeoJson functions: https://trino.io/docs/current/functions/geospatial.html#from_geojson_geometry

> ### Improvements to ANSI SQL support

Reviewer: soundex function: https://trino.io/docs/current/functions/string.html#soundex

> ### Improvements to ANSI SQL support

Reviewer: format_number function: https://trino.io/docs/current/functions/conversion.html#format_number

> ### Improvements to ANSI SQL support

Reviewer: contains function for checking if a CIDR contains an IP address: https://trino.io/docs/current/functions/ipaddress.html#ip-address-contains

SQL language support is crucial for the increasing complexities of queries and 
usage of Trino. In 2021 we added numerous new language features and 
improvements:

> * [`RANGE`]({{site.url}}/blog/2021/03/10/introducing-new-window-features.html#new%20features)

Reviewer: Also `ROWS`

> keyword for usage within a `WINDOW` function.

Reviewer: "window function"

* [`MATCH_RECOGNIZE`]({{site.url}}/blog/2021/05/19/row_pattern_matching.html)
a feature that allows for complex analysis across multiple rows. To learn more 
about this feature watch [the Community Broadcast show](/episodes/23.html).

> * Aggregate functions in for `MATCH_RECOGNIZE`.

Reviewer: This is part of MATCH_RECOGNIZE. No need to call it out separately.

* `TRUNCATE TABLE` which allows for a more efficient delete.
* `DENY` statement, which enables you to remove a user or groups access via SQL.

> * `IN catalog` clause to `CREATE ROLE`, `DROP ROLE`, `GRANT ROLE`, 

Reviewer: `IN <catalog>`

`REVOKE ROLE`, and `SET ROLE` to specify the target catalog of the statement 
instead of using the current session catalog.
* [`listagg()`]({{site.url}}/docs/current/functions/aggregate.html#listagg)
function returns concatenated values seperated by a specified separator.

> * [`SET TIME ZONE`]({{site.url}}/docs/current/sql/set-time-zone.html) keywords.

Reviewer: Describe briefly what this does: "SET TIME ZONE, to set the current time zone for the session."

* Queries that process row or array data.
* Queries that contain a `DISTINCT` clause. 
* Reduced memory usage when performing joins.

## Security

Added the following improvements and features relevant for authentication, 
authorization and integration with other security systems:

* Automatic configuration of TLS for 
[secure internal communication]({{site.url}}/docs/current/security/internal-communication.html).
This removes the need to provision per-worker TLS certificates.
* OAuth2/OIDC opaque access tokens.
* Configuring multiple password authentication plugins.
* Hiding inaccessible columns from `SELECT *` statement.
* Configuring HTTP proxy for OAuth2 authentication.
* Handling of Server Name Indication (SNI) for multiple TLS certificates.

## Data Sources

### BigQuery connector

* Added support for case insensitive name matching for BigQuery views.
* Support reading `bignumeric` type whose precision is less than or equal to 
38.
* Add support for `CREATE SCHEMA` and `DROP SCHEMA` statements.
* Improved support for BigQuery datetime and timestamp types.
* Fixed failures for queries accessing `information_schema.columns` when 
`case-insensitive-name-matching` is disabled.

> ### Cassandra connector

Reviewer 2: Support for reading user-defined types. https://github.com/trinodb/trino/issues/147

pushdown: `count`, `min`, `max`, `sum` and `avg`.
* Added support for dropping columns.
* Map ClickHouse `UUID` columns as `UUID` type in Trino instead of `VARCHAR`.

### HDFS, S3, Azure and cloud object storage systems

A core use case of Trino uses the Hive and Iceberg connectors to connect to
a data lake. These connectors differ from most as Trino is the sole query engine
as opposed to the client calling another system. Here are some changes that
for these connectors:

* Fixed queries on empty Iceberg tables without a snapshot ID that were created by Spark.
* Updated to Iceberg 0.11.0 behavior for transforms of dates and timestamps before 1970.
* Added procedure `system.flush_metadata_cache()` to flush metadata caches.
* Avoid generating splits for empty files.
* Sped up Iceberg query performance when dynamic filtering can be leveraged.
* Increased Iceberg performance when reading timestamps from Parquet files.
* Improved Iceberg performance for queries on nested data through dereference pushdown.
* Added support for `INSERT OVERWRITE` operations on S3-backed tables.
* Made the Iceberg `uuid` type available.
* Trino views made available in Iceberg.

> ### Elasticsearch connector

Reviewer 2: Better support for Elastic Cloud - https://github.com/trinodb/trino/pull/9892

* Added support for reading fields as `json` values.
* Fixed failure when documents contain fields of unsupported types.
* Added support for `scaled_float` type.
* Added support for assuming an IAM role.

> * Fixed potential incorrect results when queries contain an `IS NULL` predicate.

Reviewer: No need to list this one. Let's focus on features rather that bug fixes. (and maybe add a note somewhere about "fixed many, many, many bugs").

### MySQL connector

* Added support for reading and writing `timestamp` values with precision higher
than three.
* Added support for predicate pushdown on `timestamp` columns.
* Exclude an internal `sys` schema from schema listings.

### Pinot connector

* Updated Pinot connector to be compatible with versions >= 0.8.0 and drop 
support for older versions.
* Added support for pushdown of filters on `varbinary` columns to Pinot.
* Fixed incorrect results for queries that contain aggregations and `IN` and 
`NOT IN` filters over varchar columns.
* Fixed failure for queries with filters on `real` or `double` columns having 
`+Infinity` or `-Infinity` values.
* Implemented aggregation pushdown.
* Allowed HTTPS URLs in `pinot.controller-urls`. 

## Runtime improvements

There are a lot of performance improvements to list from the [release notes]({{site.url}}/docs/current/release.html),
Here are a few examples:

* Improved coordinator CPU utilization.
* Improved query performance by reducing CPU overhead of repartitioning data 
across worker nodes.
* Reduced graceful shutdown time for worker nodes.

## Everything else

* Added support for ARM64 in the Trino Docker image.

> * Added cancelling a Trino query execution when JDBC statement is closed.

Reviewer: This is a bug fix -- not worth mentioning.

retried. To support this intermediate stage data is persisted to replicated RAM 
or SSD.

<a title="Schokraie E, Warnken U, Hotz-Wagenblatt A, Grohme MA, Hengherr S, et al. (2012), CC BY 2.5 &lt;https://creativecommons.org/licenses/by/2.5&gt;, via Wikimedia Commons" 
 href="https://commons.wikimedia.org/wiki/File:SEM_image_of_Milnesium_tardigradum_in_active_state_-_journal.pone.0045682.g001-2.png"><img width="512" 
 alt="SEM image of Milnesium tardigradum in active state - journal.pone.0045682.g001-2" src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/cd/SEM_image_of_Milnesium_tardigradum_in_active_state_-_journal.pone.0045682.g001-2.png/512px-SEM_image_of_Milnesium_tardigradum_in_active_state_-_journal.pone.0045682.g001-2.png">
</a>

The project to introduce granular fault-tolerance into Trino is called
Project Tardigrade. It is a focus for many contributors now, and we will 
introduce you to details in the coming months. The project is named after the 
microscopic Tardigrades that are the worlds most indestructible creatures, akin
to the resiliency we are adding to Trino's queries. We look forward to telling 
you more as features unfold.

> 

Reviewer: This section makes it seem like this is the only roadmap item for 2022. We have a lot more coming, so we may want to provide a short teaser or some hint that this is just one of the many items.

---
layout: post
title:  "Trino 2021 Wrapped: A Year of Growth"
author: Brian Olsen, Martin Traverso, Manfred Moser
excerpt_separator: <!--more-->
---

As we reflect on Trino’s journey in 2021, one thing stands out. Compared to 
previous years we have seen even further accelerated, tremendous growth. Yes,
this is what all these year-in-retrospect blog posts say, but this has some 
special significance to it. This week marked the one-year anniversary since the 
project [dropped the Presto name and moved to the Trino name]({{site.url}}/blog/2020/12/27/announcing-trino.html).

## Renaming from PrestoSQL to Trino

This was a gut wrenching decision for many in the community, who had grown with
the project from early days. It also meant that the community would have to
build affinity towards the new name from scratch. We weren't sure what to expect
when we announced the rename. What followed, was an outpouring of support and
shock from the larger developer community. 

<p align="center">
 <a href="https://twitter.com/trinodb/status/1343330429684703232?s=20" target="_blank">
   <img align="center" width="100%" src="/assets/blog/2021-review/tweets.png"/>
 </a>
</p>

> <!--more-->

Author: That's just in the blog preview...The Title gets reformatted like that but any headers keep the same formatting

requests and commits started growing at a new trajectory.

<p align="center">
 <a href="https://twitter.com/bitsondatadev/status/1344028682126565381" target="_blank">
   <img align="center" width="50%" src="/assets/blog/2021-review/trending.png"/>
 </a>
</p>

At the time of writing, we just hit 4,600 stargazers on GitHub. This means, we 
have grown by over 3,000 stargazers in the last year, a 187% increase.

## Community growth

While we are on the subject, let's talk about the health of the Trino community.

Along with the growth we've seen in GitHub, we have seen a 47% growth of [the Trino Twitter](https://twitter.com/trinodb) 
followers this year. [The Trino Slack community](https://trino.io/slack.html),
where a large amount of troubleshooting and development discussions occur, saw a
75% growth, nearing 6,000 members. Finally, [the Trino YouTube channel](https://www.youtube.com/c/TrinoDB)
has seen an impressive 280% growth in subscribers. 

A lot of the increase on this channel was due to the [Trino Community Broadcast](/broadcast/), 
that brought users and contributors from the community to cover 23 episodes
about the following topics:

> * Trino ecosystem (dbt, Amundsen, Debezium, Superset) - 7

Reviewer 2: What are the numbers after hyphen? Clarify or remove.

* Trino connectors (Iceberg, Druid, Pinot) - 4
* Trino internals (Distributed Hash-Joins, Dynamic Filtering, Views) - 4
* Trino on Kubernetes (Trinetes series) - 2
* Trino users (LinkedIn, Resurface) - 2

While stargazers, subscribers, episodes, and followers tell the story of the 
growing awareness of the Trino project with the new name, what about the actual
rate of development on the project?

At the start of the year, there were 21,924 commits. This year, we pushed 3,679 
commits to the repository, sitting at over 25,600 now. Looking at the graph, this
keeps us pretty consistent with 2020's throughput.

<p align="center">
 <img align="center" width="75%" src="/assets/blog/2021-review/commits.png"/>
</p>

### 2021 by the numbers

In summary, let's take a look at the Trino project by the numbers:

* 3679 commits 💻 in GitHub 
* 3015 stargazers ⭐ in GitHub
* 2701 pull requests created ➕ in GitHub
* 2450 members 👋 in Slack
* 1979 pull requests merged ✅ in GitHub
* 1213 issues 📝 created in GitHub
* 988 followers 🐦 on Twitter
* 491 subscribers 📺 in YouTube
* 23 Trino Community Broadcast ▶️ episodes
* 17 Trino 🚀 releases
* 11 blog ✍️ posts
* 10 Trino 🍕 meetups

> * 1 Trino ⛰️ Summit

Author: Leaving this here for the historians

### Cassandra connector

* Mapped Cassandra `uuid` type to Trino `uuid`.
* Added support for Cassandra `tuple` type.
* Changed minimum number of speculative executions from two to one.
* Support for reading user-defined types.

### Clickhouse connector

* Added [ClickHouse connector]({{site.url}}/docs/current/connector/clickhouse.html).
* Improved performance of aggregation queries by computing aggregations within 
ClickHouse. Currently, the following aggregate functions are eligible for
pushdown: `count`, `min`, `max`, `sum` and `avg`.
* Added support for dropping columns.
* Map ClickHouse `UUID` columns as `UUID` type in Trino instead of `VARCHAR`.

### HDFS, S3, Azure and cloud object storage systems

A core use case of Trino uses the Hive and Iceberg connectors to connect to
a data lake. These connectors differ from most as Trino is the sole query engine
as opposed to the client calling another system. Here are some changes that
for these connectors:

* Enabled Glue statistics to support better query planning when using AWS.
* `UPDATE` support for ACID tables
* A lot of Hive view improvements.
* Parquet column indexes.

> * Target file size feature (`target_max_file_size`).

Reviewer 2: Could use better wording/detail. (I know I submitted the existing text 🙈 )
```suggestion
* `target_max_file_size` configuration to control the file size of data written by Trino.
```

* Streaming uploads to S3 by default to improve performance and reduce disk usage.
* Improved performance for tables with small files and partitioned tables.

> * Hive to Iceberg redirect.

Reviewer 2: ```suggestion
* Transparent redirection from a Hive catalog to Iceberg catalog if the table is an Iceberg table.
```

* Updated to Iceberg 0.11.0 behavior for transforms of dates and timestamps before 1970.
* Added procedure `system.flush_metadata_cache()` to flush metadata caches.
* Avoid generating splits for empty files.
* Sped up Iceberg query performance when dynamic filtering can be leveraged.
* Increased Iceberg performance when reading timestamps from Parquet files.
* Improved Iceberg performance for queries on nested data through dereference pushdown.
* Added support for `INSERT OVERWRITE` operations on S3-backed tables.
* Made the Iceberg `uuid` type available.
* Trino views made available in Iceberg.

### Elasticsearch connector

* Added support for reading fields as `json` values.
* Fixed failure when documents contain fields of unsupported types.
* Added support for `scaled_float` type.
* Added support for assuming an IAM role.
* Added retry requests with backoff when Elasticsearch is overloaded.
* Better support for Elastic Cloud.

### MemSQL connector

* Allowed limiting the size of the metadata cache.
* Added partial support for `DELETE` statement where predicate can be fully pushed 

> down to the remote datasource.

Reviewer 2: I'll still say to add a section/statement like below (and remove the MemSQL section since just a version update is not newsworthy IMO):

```
In addition to the above some more features were added that apply to most JDBC based connectors:
...
```

Reviewer: "JDBC based" is an internal implementation detail, and not something most people are familiar with. We'd need to either enumerate all the connectors or just say "to various connectors" and link to release notes or PRs for anyone who wants to get more details.

function.
* Enabled `mongodb.socket-keep-alive` config property by default.
* Add support for `json` type.
* Support reading MongoDB `DBRef` type.
* Allow skipping creation of an index for the `_schema` collection, if it 
already exists.
* Added support to redact the value of `mongodb.credentials` in the server log.
* Added support for dropping columns.

### MySQL connector

* Added support for reading and writing `timestamp` values with precision higher
than three.
* Added support for predicate pushdown on `timestamp` columns.
* Exclude an internal `sys` schema from schema listings.

### Pinot connector

* Updated Pinot connector to be compatible with versions >= 0.8.0 and drop 
support for older versions.
* Added support for pushdown of filters on `varbinary` columns to Pinot.
* Fixed incorrect results for queries that contain aggregations and `IN` and 
`NOT IN` filters over varchar columns.
* Fixed failure for queries with filters on `real` or `double` columns having 
`+Infinity` or `-Infinity` values.
* Implemented aggregation pushdown.
* Allowed HTTPS URLs in `pinot.controller-urls`. 

### Phoenix connector

* Phoenix 5 support was added.

> * Phoenix connector got improved memory usage.
> * Phoenix connector performance improved by adding ability to parallelize 
> queries within Trino 

Reviewer 2: ```suggestion
* Reduced memory usage for some queries.
* Improved performance by adding ability to parallelize 
queries within Trino.
```

or SSD.

<a title="Schokraie E, Warnken U, Hotz-Wagenblatt A, Grohme MA, Hengherr S, et al. (2012), CC BY 2.5 &lt;https://creativecommons.org/licenses/by/2.5&gt;, via Wikimedia Commons" 
 href="https://commons.wikimedia.org/wiki/File:SEM_image_of_Milnesium_tardigradum_in_active_state_-_journal.pone.0045682.g001-2.png"><img width="512" 
 alt="SEM image of Milnesium tardigradum in active state - journal.pone.0045682.g001-2" src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/cd/SEM_image_of_Milnesium_tardigradum_in_active_state_-_journal.pone.0045682.g001-2.png/512px-SEM_image_of_Milnesium_tardigradum_in_active_state_-_journal.pone.0045682.g001-2.png">
</a>

The project to introduce granular fault-tolerance into Trino is called
Project Tardigrade. It is a focus for many contributors now, and we will 
introduce you to details in the coming months. The project is named after the 
microscopic Tardigrades that are the worlds most indestructible creatures, akin
to the resiliency we are adding to Trino's queries. We look forward to telling 
you more as features unfold.

Along with Project Tardigrade will be a series of changes focused around faster

> performance in the query engine using columanar evaluation, adaptive planning,

Reviewer 2: ```suggestion
performance in the query engine using columnar evaluation, adaptive planning,
```

* `DENY` statement, which enables you to remove a user or groups access via SQL.
* `IN <catalog>` clause to `CREATE ROLE`, `DROP ROLE`, `GRANT ROLE`, 
`REVOKE ROLE`, and `SET ROLE` to specify the target catalog of the statement 
instead of using the current session catalog.

### Query processing improvements

* Added support for automatic query retries (this feature is very experimental
with some limitations for now).
* Transparent query retries.
* Updated the behavior of `ROW` to `JSON` cast to produce `JSON` objects instead
of `JSON` arrays. 
* Column and table lineage tracking in `QueryCompletedEvent`.

## Performance improvements

Improved performance for the following operations:

* Querying Parquet data for files containing column indexes.
* Reading dictionary-encoded Parquet files.
* Queries using [`rank()`]({{site.url}}/docs/current/functions/window.html#rank) window function.
* Queries using [`sum()`]({{site.url}}/docs/current/functions/aggregate.html#sum)
and [`avg()`]({{site.url}}/docs/current/functions/aggregate.html#avg) for 
decimal types.
* Queries using `GROUP BY` with single grouping column.
* Aggregation on decimal values.

> * Evaluation of the WHERE and SELECT clause.

Reviewer: Surround WHERE and SELECT with `

* Computing the product of decimal values with precision larger than 19.
* Queries that process row or array data.
* Queries that contain a `DISTINCT` clause.
* Reduced memory usage and improved performance of joins.
* `ORDER BY LIMIT` performance was improved when data was pre-sorted.
* Node-local Dynamic Filtering

## Security

Added the following improvements and features relevant for authentication, 
authorization and integration with other security systems:

* Automatic configuration of TLS for 
[secure internal communication]({{site.url}}/docs/current/security/internal-communication.html).
This removes the need to provision per-worker TLS certificates.
* Access control for materialized views.

> * OAuth2/OIDC opaque access tokens.

Reviewer: Link to OAuth2 docs: https://trino.io/docs/current/security/oauth2.html

> * Configuring multiple password authentication plugins.

Reviewer: Link to https://trino.io/docs/current/security/authentication-types.html#multiple-password-authenticators

* Hiding inaccessible columns from `SELECT *` statement.

> * Configuring HTTP proxy for OAuth2 authentication.

Reviewer: Move close to the other OAuth2 entry

> * Handling of Server Name Indication (SNI) for multiple TLS certificates.

Reviewer: Move closer to the other TLS item

that use common code. These features improve performance using:

* [Statistical aggregate function pushdown ](https://trino.io/docs/current/release/release-352.html#mysql-connector)
* [TopN pushdown and join pushdown](https://trino.io/docs/current/release/release-353.html)
* [Improved planning times by reducing number of connections opened](https://trino.io/docs/current/release/release-353.html)
* [Improved performance by improving metadata caching hit rate](https://trino.io/docs/current/release/release-356.html)
* [Rule based identifier mapping support](https://trino.io/docs/current/release/release-357.html)
* [DELETE, non-transactional inserts and write-batch-size ](https://trino.io/docs/current/release/release-360.html)
* [Metadata cache max size](https://trino.io/docs/current/release/release-361.html)
* [TRUNCATE TABLE](https://trino.io/docs/current/release/release-365.html)
* [Improved handling of Gregorian - Julian switch for date type](https://trino.io/docs/current/release/release-366.html)
* Ensured correctness when pushing down predicates and topN to remote system 
that is case-insensitive or sorts differently from Trino.

## Runtime improvements

> There are a lot of performance improvements to list from the [release notes]({{site.url}}/docs/current/release.html),

Reviewer: Sentence should end in period, not comma.

Here are a few examples:

* Improved coordinator CPU utilization.
* Improved query performance by reducing CPU overhead of repartitioning data 
across worker nodes.
* Reduced graceful shutdown time for worker nodes.

## Everything else

* [HTTP Event listener](https://trino.io/docs/current/admin/event-listeners-http.html)

> * Added support for ARM64 in the Trino Docker image.

Reviewer: Link to https://hub.docker.com/r/trinodb/trino
